Add new deploy script

mraveux opened the pull request:
This PR replace the previous release.sh script by a new ./src/deploy.sh script taking care of the release steps for you.

The script is currently untested, thus the PR and not the instant push on master.
Once the script is tested and accepted, the PR will be merged.

Edit:
The script has been tested, and used to deploy the last wallet versions on testnet & mainnet, and seems to work as expected and handle most obvious edge cases.

mraveux committed:
Replace release.sh by a new deploy.sh

mraveux committed:
deploy.sh: add some colors

sisou commented on scripts/deploy.sh:
```diff
@@ -0,0 +1,223 @@
+EXCLUDE_RELEASE=""
+SYNC_TRANSLATIONS=true
+BUILD_ENV="testnet"
+DEPLOY_ONLY=false
```
DEPLOYER probably shouldnt' have a default ;)

And I would also like BUILD_ENV not to have a default, but to always require explicitly specifying it on the command line.

mraveux replied:
that's true! I wanted to remove it and forgot, my bad! 👍 

I though about it but went for the default testnet deployment. I'll make it required/explicit then ;)

mraveux replied:
DEPLOYER is now required, with no default value.

I also added a `--testnet` parameter, and made it required to have at least either `--testnet` or `--mainnet`.

sisou commented on scripts/deploy.sh:
```diff
@@ -0,0 +1,223 @@
+
+# Commit changes
+echo -e "${CYAN}Committing changes...${NC}"
+git commit -m "$(create_message)"
```
I personally always do a `git status` after adding and before commiting, to check if it doesn't look totally off what I'm doing. E.g. if there's only a minor change in the source, I don't expect all files to have been replaced and vice versa.

So I'd like a confirm step here.

mraveux replied:
Make sense, will add that then 👍

mraveux committed:
make message and deployer required

and remove the default deployer value as well as put the message under the -m parameter for consistency

mraveux committed:
DRY the “Usage” message into a reusable function

mraveux committed:
Make BUILD_ENV required and explicit

mraveux committed:
Add “git status” and [y/N] step before commit

mraveux commented:
I applied the feedback from @sisou, lemme know if that's better for you ;)

danimoh commented on scripts/deploy.sh:
```diff
@@ -0,0 +1,274 @@
+EXCLUDE_RELEASE=""
+SYNC_TRANSLATIONS=true
+BUILD_ENV=""
+DEPLOY_ONLY=false
```
~~This overwrites the `DEPLOY_ONLY` set above.~~
You already fixed this.

danimoh commented on scripts/deploy.sh:
```diff
@@ -0,0 +1,274 @@
+    case $1 in
+        --deployer=*)
+            DEPLOYER="${1#*=}"
+            if [ $# -eq 0 ]; then
```
You are parsing the `DEPLOYER` from what follows the `=`, so the check for an additional parameter seems wrong.

mraveux replied:
that's true! I removed the unnecessary check ;)

mraveux committed:
fix —deploy-only behavior

mraveux committed:
re-enable uncommitted changes check + more verbose

mraveux commented:
The script has been tested for release v3.0.11 and seems to work.
I will still add some error handling, in case a command fail so that the script doesn't continue and mess it up.

Except that, the script is considered done, until further feedbacks.

mraveux committed:
Add error handling

mraveux commented:
Just pushed some error handling.
Script is considered done for now and ready for final review.

danimoh commented on scripts/deploy.sh:
```diff
@@ -0,0 +1,292 @@
+# Validate version number format
+if ! [[ $VERSION =~ ^[0-9]+\.[0-9]+\.[0-9]+$ ]] && [ "$DEPLOY_ONLY" = false ]; then
+    show_usage "Version number must be in format X.Y.Z"
+fi
```
This check could move into the `if` block above, which already checks for `[ "$DEPLOY_ONLY" = false ];`

danimoh commented on scripts/deploy.sh:
```diff
@@ -0,0 +1,292 @@
+        fi
+    done
+
+    echo -e "${GREEN}Deployment complete!${NC}"
```
This gives a false sense of safety that everything succeeded.
Maybe print that the user should still look through the logs.

danimoh commented on scripts/deploy.sh:
```diff
@@ -0,0 +1,292 @@
+    echo "Nimiq Wallet v$VERSION
+
+$COMMIT_MSG
+
+$EXCLUDE_RELEASE"
```
I guess this leaves a trailing new line when `$EXCLUDE_RELEASE` is empty?

danimoh commented on scripts/deploy.sh:
```diff
@@ -0,0 +1,292 @@
+    fi
+done
+
+# Copy build files
```
Should run the `./clean.sh` script here before copying the files.

mraveux committed:
Add testnet servers, and make build_env required

danimoh reviewed (approved):
Very cool, thank you for this :+1:
The issues I showed you in CR, I see you already addressed.
Commenting on some additional things, but you don't necessarily need to address them, if you don't want to.

mraveux committed:
Add a confirmation “recap” before running script

mraveux committed:
Add —same-as to allow to deploy same version

For example “—same-as=testnet —mainnet” will deploy the version already deployed on the testnet to the mainnet

mraveux committed:
add ./clean.sh execution before copying dist files

mraveux committed:
prevent trailing newline if no —exclude-release

danimoh commented on scripts/deploy.sh:
```diff
@@ -57,11 +60,23 @@ DEPLOY_SERVERS=()
+elif [[ "$1" =~ ^--.*$ ]]; then
+    # If first argument is any other flag
+    VERSION=""
+    shift
```
This discards the flag without any warning. Better would probably be to exit with an error.

mraveux committed:
Add post-deployment message

mraveux committed:
Remove unnecessary —deployer check

mraveux committed:
Forbid version if —deploy-only or —same-as
